Add __call__ wrappers to aiosignal along with better type hinting extensions

Vizonex opened the pull request:
<!-- Thank you for your contribution! -->

## What do these changes do?

I'm going to start moving some of my code over from aiocallback over to this branch which I think would be acceptable given the circumstances of this [proposal](https://github.com/aio-libs/aiohttp/issues/11036#issuecomment-2912504991) it was concluded that moving my things over here would allow this library to hopefully achieve a better user development experience.

Eventually I will give this library special member descriptors so that much of the works done with TraceConfig can be less bulky overall but this is the start to doing so.

## Are there changes in behavior for the user?

<!-- Outline any notable behaviour for the end users. -->

## Related issue number

<!-- Are there any issues opened that will be resolved by merging this change? -->

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

Vizonex commented:
Note: This pull request was my backup plan if things went extremely poorly. I don't expect it to happen at all.

Dreamsorcerer commented on aiosignal/__init__.pyi:
Not sure why we have typing stubs here. Should make another PR later to move the annotations inline.

Dreamsorcerer commented on aiosignal/typedefs.py:
I'd just put this at the top of the file, I'm not sure we'll need a separate file to manage typing aliases currently.

Vizonex commented:
@Dreamsorcerer I removed the stub file as well and made a few suggested changes that you wanted. I think I will transform aiocallback into a member descriptor library instead so that I don't have to deprecate my end but I think once these features are added in again I can re-add aiosignal back to my library, originally I dropped usage of this library when I saw an issue that this library would be deprecated.

Dreamsorcerer commented on aiosignal/__init__.py:
```diff
@@ -16,21 +29,26 @@ class Signal(FrozenList):
     __slots__ = ("_owner",)
 
-    def __init__(self, owner):
+    def __init__(self, owner:_O):
```
The owner parameter doesn't seem to have any actual use, so I don't see any reason to track it with a TypeVar:
```suggestion
    def __init__(self, owner: object):
```

Vizonex replied:
@Dreamsorcerer Makes sense. I just didn't know if users wanted to remember what objects were being carried around or not.

Vizonex commented:
> It was a proposal, given that the library was basically 3 lines of code. You've given us a counter-proposal to expand the library into something worth keeping around.
@Dreamsorcerer Glad I was able to do so. I didn't want to see something this creative being put to waste. Especially since I was inspired by a few legendary libraries like discord-py and fastapi to keep it going and try and expand upon the idea that users could have a creative experience with very minimal code. 

I did have the idea of creating a subclass signal that carries the owner in the front But maybe I'll save that for a future pull request.

Vizonex commented:
@Dreamsorcerer I wanted to ask but could I make the owner variable optional and let it's default be None? There might be a use-case where a user might not want to pass an owner variable and instead use the signal as a wrapper and sender without a class of any sort.

Dreamsorcerer commented:
> @Dreamsorcerer I wanted to ask but could I make the owner variable optional and let it's default be None? There might be a use-case where a user might not want to pass an owner variable and instead use the signal as a wrapper and sender without a class of any sort.

I guess, do it a separate PR though after this one. I thought the parameter was pointless anyway.

Dreamsorcerer commented on CHANGES/699.feature:
Apparently not:
```
Check "CHANGES" folder... 
/code/CHANGES/699.feature.1.rst has illegal suffix
/code/CHANGES/699.feature.2.rst has illegal suffix

Allowed suffixes are: ['.feature', '.bugfix', '.doc', '.removal', '.misc', '.feature.rst', '.bugfix.rst', '.doc.rst', '.removal.rst', '.misc.rst']
```

webknjaz commented on aiosignal/__init__.py:
```diff
@@ -1,11 +1,23 @@
+else:
+    from typing_extensions import ParamSpec
+
+P = ParamSpec("P")
+T = typing.TypeVar("T")
+AsyncFunc = typing.Callable[P, typing.Awaitable[T]]
```
Can we make these names have a single leading underscore?

Vizonex replied:
Sure. Give me a second and I'll change that real quickly.

webknjaz replied:
Marking this as resolved.

webknjaz commented on CHANGES/699.feature.2.rst:
```diff
@@ -0,0 +1 @@
+Added optional library typing-extensions if Python 3.9 is used. -- by :user Vizonex:
```
The role syntax is invalid. Also, we still don't have `:user:` so it wouldn't be interpreted unless `sphinx-issues` is integrated. Additionally, I meant that the trailing period is put after the byline if it's a part of the sentence. Hence the suggested "inline code" syntax for now:
```suggestion
Added optional library ``typing-extensions`` under Python 3.9 runtime or older -- by ``@Vizonex``.
```

webknjaz commented on aiosignal/__init__.py:
```diff
@@ -1,11 +1,23 @@
+if sys.version_info >= (3, 10):
+    from typing import ParamSpec
+else:
+    from typing_extensions import ParamSpec
```
This requires a conditional runtime dependency declared in `setup.cfg`

webknjaz commented on aiosignal/__init__.py:
```diff
@@ -16,21 +28,26 @@ class Signal(FrozenList):
+            await receiver(*args, **kwargs)
+
+    def __call__(self, func: AsyncFunc[_P, _T]) -> AsyncFunc[_P, _T]:
+        """wraps a callback function to the signal."""
```
Could you use an imperative mood per PEP 257?

Also, I don't understand what “wrapping” refers to. It seems like the implementation appends callbacks w/o wrapping anything. Is it a reference to a decorator use? Any way to rephrase this?

Vizonex replied:
Yeah I can rephrase the documentation, but it basically is a decorator for appending functions.

Dreamsorcerer replied:
I was going to clean this up once the test pass. These just mean it's a decorator (as evidenced in the test).

webknjaz commented on CHANGES/699.feature.1.rst:
```diff
@@ -0,0 +1 @@
+Added ParamSpec, Generic Variables and `__call__` functions to aiosignal & Removed the stubfile. -- by :user Vizonex:
```
Oh, I mean whether you're still fixing the change note or are ready to receive a new round of feedback on things that are broken in this variant.

webknjaz replied:
Yeah, but this should still be formatted into a coherent paragraph with refs et al.

webknjaz commented on setup.cfg:
```diff
@@ -37,7 +37,9 @@ packages = find:
 install_requires =
-  frozenlist >= 1.1.0
+  frozenlist >= 1.1.0,
+  typing_extensions; python_version == '3.9'
```
What's the minimum version of the dependency implementing `ParamSpec`? We should set the lower limit to that.

Additionally, please use `-` instead of `_` in the PyPI project name on this line.

webknjaz commented:
@Vizonex could you set `fail_under = 98` in `.coveragerc`? Add a code comment explaining that it's a temporary hack until one or more of `covdefaults`, `coverage-simple-excludes` and `coverage-conditional-plugin` is integrated?

webknjaz commented on setup.cfg:
```diff
@@ -37,7 +37,9 @@ packages = find:
 install_requires =
-  frozenlist >= 1.1.0
+  frozenlist >= 1.1.0,
+  typing_extensions; python-version == '3.9'
```
(the PyPI project name comes before the semicolon, not after)
```suggestion
  typing-extensions; python_version == '3.9'
```

Vizonex commented:
> I'm planning to address the changelog+coverage-related config differences before considering merging this PR. Hope it can be rebased on top of those first.

@webknjaz Good Idea, hopefully after this release maybe I can work on fixing aiosignal for aiohttp's TraceConfig when were done. And maybe provide some updated documentation for aiosignal afterwards.

Vizonex commented:
@webknjaz I'm going to rename the features files by move the number to 699.1.feature.rst so that the pre-commit.ci finds it to be acceptable...

Vizonex commented:
Good news. It passes

webknjaz reviewed (changes requested):
Putting reordering components in fragment names causes makes them invalid.

Vizonex commented:
I remerge both Feature commits because I don't like that this workflow will not pass we might be able to fix this afterwards but for right now. This got beyond annoying.

Vizonex commented:
@Dreamsorcerer Thanks for getting this pull request passed I appreciate it. Let me know when the update comes out so I can re-add this library back to my member descriptor library.

Dreamsorcerer commented:
> Let me know when the update comes out

We won't release it without addressing my previous comment first, and getting the strict typing working.

Dreamsorcerer commented:
I think we do need 2 implementations or something here.

One with TypeVarTuple for Callable support, and another for ParamSpec.
```
def send(self, *args: *Ts) -> T:
def send(self, *args: P.args, **kwargs: P.kwargs) -> T:
```

Or, given that this is mostly going to be handling callbacks, maybe we could just do the first one and not support kwargs...
